proc_macro: Improve Debug representations

alexcrichton opened the pull request:
This commit improves the `fmt::Debug` output of `proc_macro` data structures by
primarily focusing on the representation exposed by `proc_macro` rather than the
compiler's own internal representation. This cuts down quite a bit on assorted
wrapper types and ensure a relatively clean output.

Closes #49720

alexcrichton committed:
proc_macro: Improve Debug representations

This commit improves the `fmt::Debug` output of `proc_macro` data structures by
primarily focusing on the representation exposed by `proc_macro` rather than the
compiler's own internal representation. This cuts down quite a bit on assorted
wrapper types and ensure a relatively clean output.

Closes #49720

dtolnay commented on src/libproc_macro/lib.rs:
```diff
@@ -334,6 +335,16 @@ impl Span {
+#[unstable(feature = "proc_macro", issue = "38356")]
+impl fmt::Debug for Span {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{:?} bytes({}...{})",
```
I believe the lo/hi interval is half-open so `{}..{}` would be more accurate. Like in #49720 the `+` token is `bytes(82..83)`.

dtolnay committed:
Print proc_macro spans as a half-open range

A span covering a single byte, such as for an operator `+` token, should
print as e.g. `80..81` rather than `80...81`. The lo end of the range is
inclusive and the hi end is exclusive.

dtolnay commented:
Here is sample debug output for the tokens `[a + 1]`. This is a huge improvement from what it looked like before this PR. Thanks!

```rust
TokenStream [
    Group {
        delimiter: Bracket,
        stream: TokenStream [
            Term {
                sym: a,
                span: #0 bytes(80..81)
            },
            Op {
                op: '+',
                spacing: Alone,
                span: #0 bytes(82..83)
            },
            Literal {
                lit: Integer(
                    1
                ),
                suffix: None,
                span: #0 bytes(84..85)
            }
        ],
        span: #0 bytes(79..86)
    }
]
```

@bors r+

bors commented:
:pushpin: Commit 52766b5 has been approved by `dtolnay`

<!-- @bors r=dtolnay 52766b57477002583b0a72c917a172c082cb4a36 -->

eddyb commented on src/libproc_macro/lib.rs:
```diff
@@ -717,7 +742,8 @@ impl fmt::Display for Term {
 pub struct Literal {
-    token: token::Token,
+    lit: token::Lit,
+    suffix: Option<ast::Name>,
```
Oh, nice!
